src/io/utils.rs

```diff
@@ -200,34 +204,118 @@ where
+		log_error!(logger, "Failed to deserialize OutputSweeper: {}", e);
+		std::io::Error::new(std::io::ErrorKind::InvalidData, "Failed to deserialize OutputSweeper")
+	})
+}
+
+/// Read previously persisted spendable output information from the store and migrate to the
+/// upstreamed `OutputSweeper`.
+///
+/// We first iterate all `DeprecatedSpendableOutputInfo`s and have them tracked by the new
+/// `OutputSweeper`. In order to be certain the initial output spends will happen in a single
+/// transaction (and safe on-chain fees), we batch them to happen at current height plus two
+/// blocks. Lastly, we remove the previously persisted data once we checked they are tracked and
+/// awaiting their initial spend at the correct height.
+///
+/// Note that this migration will be run in the `Builder`, i.e., at the time when the migration is
+/// happening no background sync is ongoing, so we shouldn't have a risk of interleaving block
+/// connections during the migration.
+pub(crate) fn migrate_deprecated_spendable_outputs<K: KVStore + Sync + Send, L: Deref>(
+	sweeper: Arc<Sweeper<K>>, kv_store: Arc<K>, logger: L,
+) -> Result<(), std::io::Error>
 where
 	L::Target: Logger,
 {
-	let mut res = Vec::new();
+	let best_block = sweeper.current_best_block();
 
 	for stored_key in kv_store.list(
-		SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
-		SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+		DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+		DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
 	)? {
 		let mut reader = Cursor::new(kv_store.read(
-			SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
-			SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+			DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+			DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
 			&stored_key,
 		)?);
-		let output = SpendableOutputInfo::read(&mut reader).map_err(|e| {
+		let output = DeprecatedSpendableOutputInfo::read(&mut reader).map_err(|e| {
 			log_error!(logger, "Failed to deserialize SpendableOutputInfo: {}", e);
 			std::io::Error::new(
 				std::io::ErrorKind::InvalidData,
 				"Failed to deserialize SpendableOutputInfo",
 			)
 		})?;
-		res.push(output);
+		let descriptors = vec![output.descriptor.clone()];
+		let spend_delay = Some(best_block.height + 2);
+		sweeper.track_spendable_outputs(descriptors, output.channel_id, true, spend_delay);
+		if let Some(tracked_spendable_output) =
+			sweeper.tracked_spendable_outputs().iter().find(|o| o.descriptor == output.descriptor)
+		{
+			match tracked_spendable_output.status {
+				OutputSpendStatus::PendingInitialBroadcast { delayed_until_height } => {
+					if delayed_until_height == spend_delay {
+						kv_store.remove(
+							DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_PRIMARY_NAMESPACE,
+							DEPRECATED_SPENDABLE_OUTPUT_INFO_PERSISTENCE_SECONDARY_NAMESPACE,
+							&stored_key,
+							false,
+						)?;
```

Comment:
Should we persist the `Sweeper` before removing these?

Reply:
Oh, never mind. It looks like `track_spendable_outputs` does just that.

Reply:
Though it will only log but otherwise fail silently. Is this a concern?

Reply:
Feel free to squash. We can update this in a follow-up as soon as a new RC is available.